test/utils/Decimal.ts

```diff
@@ -1,7 +1,201 @@
 import { ethers } from "hardhat";
 import { BigNumber } from "ethers";
 
-export type Numberish = Number | string;
+export type Numberish = Number | number | string | BigInt | BigNumber | Decimal;
+
+/**
+ * Matches the most common ERC20 18-decimals precision, such as wETH
+ */
+export const DEFAULT_DECIMAL_PRECISION = 18;
+
+ /**
+  * Creates a new `Decimal` Fixed-Point Decimal type
+  * The default precision is 18 decimals. Any excess digits will be truncated.
+  * @param value Number-like value to convert to Decimal
+  * @param decimals Decimals precision after the fraction, excess is truncated
+  */
+export function decimal(value:Numberish, decimals:number = DEFAULT_DECIMAL_PRECISION): Decimal {
+  return new Decimal(value, decimals);
+}
+
+/**
+ * @abstract A Fixed-Point Decimal type with a strongly defined `decimals` precision
+ *           compatible with ERC20.decimals() concept.
+ */
+export class Decimal {
+  readonly int: bigint; // big integer that holds the FixedPoint Decimal value
+  readonly decimals: number; // number of decimal digits that form a fraction, can be 0
+
+  constructor(value:Numberish, decimals:number) {
+    this.decimals = decimals;
+    this.int = Decimal.toFixedPointInteger(value, decimals);
+    Object.freeze(this);
+  }
+
+  /** @returns BigNumber from this Decimal */
+  public toBigNumber(): BigNumber {
+    return BigNumber.from(this.int.toString());
```

Comment:
According to ethers.js docs this support bigint as an input, i.e. 
```suggestion
    return BigNumber.from(this.int);
```

Did you try that?

Reply:
Yeah, indeed I tried that. I checked their source code and the reason I chose `toString()` there was to just make it faster.
Their implementation for `bigint` support is recursive:
```ts
        if (typeof(anyValue) === "bigint") {
            return BigNumber.from(anyValue.toString());
        }
```
So passing `string` is actually the most straightforward way. Everything is kept as a string internally for `BigNumber`

Reply:
JavaScript 😢